Intacct: Added churros for missing endpoints

MadhuriKadam-GSLab opened the pull request:
## Highlights
* Added churros for credit-memos, debit-memos, deposits and sales-receipts APIs 

## Non-Customer Highlights : 
* For contracts object we dont have permission hence it is failing
*  Sales-order API is failing on snapshot and staging also, not have correct values in the payload
## Screenshot
> Churros initialization successful!
PS C:\Users\GS-1259\CE-4656\churros\finalch\churros-sauce> churros test elements/intacct


info: Using url: http://localhost:8080
info: Using user: madhuri.kadam@gslab.com
info: Running tests for element: intacct
info: Provisioned with instance id of 817
  Basic tests
    √ should provision
    √ should GET /objects (457ms)
    - docs
    √ metadata (489ms)
error: Failed to retrieve or validate: /hubs/finance/attachments
error: Failed to retrieve or validate: /hubs/finance/contracts
    1) transformations
    - should not allow provisioning with bad credentials

  bills-payments
    √ should allow SR for /hubs/finance/bills-payments (7145ms)
    √ should support Ceql date search (3596ms)
    √ should allow paginating with page and nextPage /hubs/finance/bills-payments (4569ms)

  bills
    √ should allow CRDS for /hubs/finance/bills (19102ms)
    √ should allow paginating with page and pageSize for /hubs/finance/bills (7654ms)
    √ should support updated > {date} Ceql search (5123ms)

  chargecard-transactions
    √ should allow SR for /hubs/finance/chargecard-transactions (3777ms)
    √ should allow paginating with page and nextPage /hubs/finance/chargecard-transactions (4129ms)
    √ should support Ceql WHENMODIFIED search (1294ms)

  checking-accounts
    √ should allow SR for /hubs/finance/checking-accounts (3665ms)
    √ should allow paginating with page and nextPage /hubs/finance/checking-accounts (5062ms)
    √ should support Ceql WHENMODIFIED search (2487ms)

  classes
    √ should allow CRUDS for /hubs/finance/classes (11444ms)
    √ should allow paginating with page and pageSize for /hubs/finance/classes (6253ms)
    √ should support updated > {date} Ceql search (2244ms)

  contacts
    √ should allow CRDS for /hubs/finance/contacts (8842ms)
    √ should allow PATCH for /hubs/finance/contacts/{id} (4717ms)
    √ should allow paginating with page and pageSize for /hubs/finance/contacts (5414ms)
    √ should support status Ceql search (3051ms)

  contracts
error: Failed to create or validate: /hubs/payment/contracts
    2) should allow CRUDS for /hubs/payment/contracts
error: Failed to retrieve or validate: /hubs/payment/contracts
    3) should allow paginating with page and nextPage /hubs/payment/contracts

  credit-memos
    √ should allow CRS for /hubs/finance/credit-memos (5637ms)
    √ should allow paginating with page and pageSize for /hubs/finance/credit-memos (3873ms)
    √ should support customerid=AGR001 Ceql search (3223ms)

  customers
    √ should allow CRUDS for /hubs/finance/customers (11421ms)
    √ should allow paginating with page and pageSize for /hubs/finance/customers (7151ms)
    √ should support updated > {date} Ceql search (3009ms)

  debit-memos
    √ should allow CRS for /hubs/finance/debit-memos (6117ms)
    √ should allow paginating with page and pageSize for /hubs/finance/debit-memos (4689ms)
    √ should support vendorid= V102 Ceql search (949ms)

  departments
    √ should allow CRUDS for /hubs/finance/departments (8181ms)
    √ should allow paginating with page and pageSize for /hubs/finance/departments (6411ms)
    √ should support updated > {date} Ceql search (2474ms)

  deposits
    √ should allow SR for /hubs/finance/deposits (2211ms)
    √ should allow CRUDS for /hubs/finance/deposits (5848ms)
    √ should allow paginating with page and nextPage /hubs/finance/deposits (2747ms)
    √ should support Ceql WHENMODIFIED search (1880ms)

  employees
    √ should allow CRDS for /hubs/finance/employees (12209ms)
    √ should allow PATCH for /hubs/finance/employees/{id} (11945ms)
    √ should allow paginating with page and pageSize for /hubs/finance/employees (8154ms)
    √ should support updated > {date} Ceql search (4289ms)

  exchange-rate-entries
    √ should allow SR for /hubs/payment/exchange-rate-entries (3604ms)
    √ should allow paginating with page and nextPage /hubs/payment/exchange-rate-entries (2831ms)

  exchange-rate-types
    √ should allow SR for /hubs/payment/exchange-rate-types (3369ms)
    √ should allow paginating with page and nextPage /hubs/payment/exchange-rate-types (2929ms)

  exchange-rates
    - should allow CRS for /hubs/payment/exchange-rates
    - should allow paginating with page and nextPage /hubs/payment/exchange-rates

  expense-reports
    √ should allow CRDS for /hubs/finance/expense-reports (11275ms)
    √ should allow PATCH for /hubs/finance/expense-reports/{id} (7182ms)
    √ should allow paginating with page and pageSize for /hubs/finance/expense-reports (6682ms)
    √ should support Ceql batchKey search (2553ms)

  folders
    √ should allow CRUDS for /hubs/finance/folders (12646ms)
    √ should allow paginating with page and pageSize for /hubs/finance/folders (5125ms)
    √ should support description = {string} Ceql search (1744ms)
    √ should allow CRUDS for /hubs/finance/folders/5agy3oozuw/files (6322ms)
    √ should allow paginating with page and pageSize for /hubs/finance/folders/2hk3c3y1rp/files (5841ms)
    √ should support  description = {string} Ceql search (1217ms)

  invoices
error: Failed to delete /hubs/finance/folders/5agy3oozuw
    √ should allow CRUDS for /hubs/finance/invoices (16084ms)
    √ should allow paginating with page and pageSize for /hubs/finance/invoices (6263ms)
    √ should support updated > {date} Ceql search (3625ms)

  items
    √ should allow CRUDS for /hubs/finance/items (14300ms)
    √ should allow paginating with page and pageSize for /hubs/finance/items (5841ms)
    √ should support updated > {date} Ceql search (3089ms)

  journals
    √ should allow CRUDS for /hubs/finance/journals (9698ms)
    √ should allow paginating with page and pageSize for /hubs/finance/journals (6002ms)
    √ should support updated > {date} Ceql search (2145ms)

  ledger-accounts
    √ should allow CRUDS for /hubs/finance/ledger-accounts (11893ms)
    √ should allow paginating with page and pageSize for /hubs/finance/ledger-accounts (5879ms)
    √ should support updated > {date} Ceql search (2190ms)

  locations
    √ should allow CRUDS for /hubs/finance/locations (10954ms)
    √ should allow paginating with page and pageSize for /hubs/finance/locations (7546ms)
    √ should support updated > {date} Ceql search (2726ms)

  /objects
    √ should support GET /objects/chargecardTransactions/metadata (2142ms)
    √ should support GET /objects/chargecardTransactions/metadata customFieldsOnly parameter (2883ms)
    √ should support GET /objects/billsPayments/metadata (3963ms)
    √ should support GET /objects/billsPayments/metadata customFieldsOnly parameter (2283ms)
    √ should support GET /objects/checkingAccounts/metadata (2550ms)
    √ should support GET /objects/checkingAccounts/metadata customFieldsOnly parameter (2504ms)
    √ should support GET /objects/exchangeRateEntries/metadata (1358ms)
    √ should support GET /objects/exchangeRateEntries/metadata customFieldsOnly parameter (2159ms)
    √ should support GET /objects/exchangeRates/metadata (1196ms)
    √ should support GET /objects/exchangeRates/metadata customFieldsOnly parameter (2420ms)
    √ should support GET /objects/exchangeRateTypes/metadata (1381ms)
    √ should support GET /objects/exchangeRateTypes/metadata customFieldsOnly parameter (1890ms)

  payments
    - should allow CRS for /hubs/finance/payments
    - should allow paginating with page and pageSize for /hubs/finance/payments
    - should support updated > {date} Ceql search
    - should allow C for /hubs/finance/payments/{id}/apply

  ping
    √ should allow GET for /hubs/finance/ping (236ms)

  polling
    - polling /hubs/finance/customers
    - polling /hubs/finance/employees
    - polling /hubs/finance/invoices
    - polling /hubs/finance/items
    - polling /hubs/finance/journals
    - polling /hubs/finance/ledger-accounts
    - polling /hubs/finance/payments
    - polling /hubs/finance/vendors

  projects
    √ should allow CRUDS for /hubs/finance/projects (11786ms)
    √ should allow paginating with page and pageSize for /hubs/finance/projects (6152ms)
    √ should support updated > {date} Ceql search (2971ms)

  purchase-orders
    - should allow CRDS for /hubs/finance/purchase-orders
    - should allow paginating with page and pageSize for /hubs/finance/purchase-orders
    - should support updated > {date} Ceql search

  reporting-periods
    √ should allow GET for /hubs/finance/reporting-periods (2713ms)
    √ should allow paginating with page and pageSize for /hubs/finance/reporting-periods (7372ms)
    √ should support status = {string} Ceql search (2868ms)

  sales-orders
error: Failed to create or validate: /hubs/finance/sales-orders
    4) should allow CRUDS for /hubs/finance/sales-orders
    √ should allow paginating with page and pageSize for /hubs/finance/sales-orders (7120ms)
    √ should support updated > {date} Ceql search (4491ms)

  sales-receipts
    √ should allow CRS for /hubs/finance/sales-receipts (9953ms)
    √ should allow paginating with page and nextPage /hubs/finance/sales-receipts (3791ms)
    √ should support Ceql WHENMODIFIED search (1569ms)

  terms
    √ should allow SR for /hubs/finance/terms (3893ms)
    √ should allow CUD for /hubs/finance/terms (5984ms)
    √ should allow paginating with page and pageSize for /hubs/finance/terms (6844ms)
    √ should support updated > {date} Ceql search (2025ms)

  transactions
    - should allow CRDS for /hubs/finance/transactions
    - should allow GET for /hubs/finance/transactions-entries
    - should allow paginating with page and pageSize for /hubs/finance/transactions
    - should support updated > {date} Ceql search

  vendors
    √ should allow CRDS for /hubs/finance/vendors (12215ms)
    √ should allow paginating with page and pageSize for /hubs/finance/vendors (6682ms)
    √ should support updated > {date} Ceql search (4708ms)
    √ should support paginated search and return less than requested (20880ms)

  vouchers
    √ should allow CRDS for /hubs/finance/vouchers (16326ms)
    √ should allow paginating with page and pageSize for /hubs/finance/vouchers (5662ms)
    √ should support updated > {date} Ceql search (3089ms)

  warehouses
    √ should allow SR for /hubs/finance/warehouses (4002ms)
    √ should allow paginating with page and pageSize for /hubs/finance/warehouses (7381ms)
    √ should support updated > {date} Ceql search (2089ms)


  109 passing (12m)
  23 pending
  4 failing

  1) Basic tests transformations:
     Error: AssertionError: expected 200 to equal 0
      at cloud.delete.catch.then.catch.then.then.catch.then (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\test\e
lements\assets\basics.js:124:17)
      at _fulfilled (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:854:54)
      at self.promiseDispatch.done (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:883:30)
      at Promise.promise.promiseDispatch (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:816:13)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:624:44
      at runSingle (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:137:13)
      at flush (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)

  2) contracts should allow CRUDS for /hubs/payment/contracts:

      AssertionError: expected status code 400 to equal 200 - { "requestId": "5ae6dfb40ab864963e29d143", "message": "Request failed", "providerMessage": "<?xml version=\"1.0\" encoding=\"UTF-8\"?><response><control><status>success</status><senderid>Cloud Elements</senderid><controlid>EFDWnk2vPu</controlid><uniqueid>false</uniqueid><dtdversion>3.0</dtdversion></control><operation><authentication><status>success</status><userid>admin</userid><companyid>Cloud Elements-DEV</companyid><sessiontimestamp>2018-04-30T02:27:53-07:00</sessiontimestamp></authentication><result><status>failure</status><function>readByQuery</function><controlid>6459906</controlid><errormessage><error><errorno>PL04000005</errorno><description></description><description2>You do not have permission for API operation READ on objects of type contract</description2><correction></correction></error></errormessage></result></operation></response>" }
      + expected - actual

      -false
      +true

      at chakram.addMethod (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\assertions.js:17:3)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chakram\lib\plugins.js:123:22)
      at ctx.(anonymous function) (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\lib\chai\utils\add
Method.js:41:25)
      at doAsserterAsyncAndAddThen (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\c
hai-as-promised.js:296:33)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\chai-as-promis
ed.js:255:21)
      at ctx.(anonymous function) (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\lib\chai\utils\ove
rwriteMethod.js:49:33)
      at doAsserterAsyncAndAddThen (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\c
hai-as-promised.js:296:33)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\chai-as-promis
ed.js:255:21)
      at ctx.(anonymous function) [as statusCode] (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\li
b\chai\utils\overwriteMethod.js:49:33)
      at run (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\suite.js:735:133)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\cloud.js:25:7
      at chakram.post.then.r (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\cloud.js:53:44)
      at _fulfilled (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:854:54)
      at self.promiseDispatch.done (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:883:30)
      at Promise.promise.promiseDispatch (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:816:13)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:624:44
      at runSingle (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:137:13)
      at flush (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)

  3) contracts should allow paginating with page and nextPage /hubs/payment/contracts:

      AssertionError: expected status code 400 to equal 200 -  
{ "requestId": "5ae6dfb40ab864963e29d143", "message": "Request failed", "providerMessage": "<?xml version=\"1.0\" encoding=\"UTF-8\"?><response><control><status>success</status><senderid>Cloud Elements</senderid><controlid>EFDWnk2vPu</controlid><uniqueid>false</uniqueid><dtdversion>3.0</dtdversion></control><operation><authentication><status>success</status><userid>admin</userid><companyid>Cloud Elements-DEV</companyid><sessiontimestamp>2018-04-30T02:27:53-07:00</sessiontimestamp></authentication><result><status>failure</status><function>readByQuery</function><controlid>6459906</controlid><errormessage><error><errorno>PL04000005</errorno><description></description><description2>You do not have permission for API operation READ on objects of type contract</description2><correction></correction></error></errormessage></result></operation></response>" }
      + expected - actual

      -false
      +true

      at chakram.addMethod (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\assertions.js:17:3)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chakram\lib\plugins.js:123:22)
      at ctx.(anonymous function) (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\lib\chai\utils\add
Method.js:41:25)
      at doAsserterAsyncAndAddThen (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\c
hai-as-promised.js:296:33)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\chai-as-promis
ed.js:255:21)
      at ctx.(anonymous function) (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\lib\chai\utils\ove
rwriteMethod.js:49:33)
      at doAsserterAsyncAndAddThen (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\c
hai-as-promised.js:296:33)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\chai-as-promis
ed.js:255:21)
      at ctx.(anonymous function) [as statusCode] (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\li
b\chai\utils\overwriteMethod.js:49:33)
      at run (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\suite.js:735:133)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\cloud.js:25:7
      at chakram.get.then.r (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\cloud.js:69:44)
      at _fulfilled (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:854:54)
      at self.promiseDispatch.done (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:883:30)
      at Promise.promise.promiseDispatch (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:816:13)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:624:44
      at runSingle (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:137:13)
      at flush (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)

  4) sales-orders should allow CRUDS for /hubs/finance/sales-orders:

      AssertionError: expected status code 400 to equal 200 - {"requestId":"5adf03e7cfe2b74f4afee1fe","message":"Request
 failed","providerMessage":"error - {errorno=PL03000004, description=null, description2=You are not using the default fa
ir value price list. This is not allowed when you choose not to override fair value price list in Order Entry Setup., co
rrection=null}"}
      + expected - actual

      -false
      +true

      at chakram.addMethod (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\assertions.js:17:3)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chakram\lib\plugins.js:123:22)
      at ctx.(anonymous function) (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\lib\chai\utils\add
Method.js:41:25)
      at doAsserterAsyncAndAddThen (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\c
hai-as-promised.js:296:33)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\chai-as-promis
ed.js:255:21)
      at ctx.(anonymous function) (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\lib\chai\utils\ove
rwriteMethod.js:49:33)
      at doAsserterAsyncAndAddThen (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\c
hai-as-promised.js:296:33)
      at .<anonymous> (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai-as-promised\lib\chai-as-promis
ed.js:255:21)
      at ctx.(anonymous function) [as statusCode] (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\chai\li
b\chai\utils\overwriteMethod.js:49:33)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\cloud.js:33:25
      at chakram.post.then.r (C:\Users\GS-1259\CE-4656\churros\finalch\churros\src\core\cloud.js:53:44)
      at _fulfilled (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:854:54)
      at self.promiseDispatch.done (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:883:30)
      at Promise.promise.promiseDispatch (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:816:13)
      at C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:624:44
      at runSingle (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:137:13)
      at flush (C:\Users\GS-1259\CE-4656\churros\finalch\churros\node_modules\q\q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)





## Reference
* [SOBA](https://github.com/cloud-elements/soba/pull/8556)
* [RALLY-#${TA14099}](https://rally1.rallydev.com/#/144349237612d/detail/task/215171706340)

## Closes
* Closes [#${US10844}](https://rally1.rallydev.com/#/144349237612d/detail/userstory/210886375752)

tyler-cloud-elements commented:
@MadhuriKadam-GSLab IMO for `contracts` we should improve the err msg if we can from soba side (maybe a defect needed). If we don't have permissions - it might be worth skipping (or at least adding a comment in the test so no one ends up thinking it's an error looking into).

What's the deal with `sales-orders` ?
<img width="949" alt="screen shot 2018-04-27 at 9 28 13 pm" src="https://user-images.githubusercontent.com/13838430/39390954-ec46281e-4a61-11e8-9e33-6d2d84ce67b5.png">

MadhuriKadam-GSLab commented:
@tyler-cloud-elements  
Fixed the contract API issue. Now it is giving the proper error message. Also added the comment in the contracts.js. 

For sales-order, I did try to POST API from snapshot there also getting the same value also changed payload values still getting the same error.
